cajita/src/Cajita_SparseIndexSpace.hpp

```diff
@@ -0,0 +1,525 @@
+        uldata >>= lz;
+        ulresult <<= 1;
+        count++;
+        ulresult |= ( uldata & 1 );
+        uldata >>= 1;
+        rmask <<= lz + 1;
+    }
+    ulresult <<= 64 - count; // 64 bits (maybe not use a constant 64 ...?)
+    ulresult = binary_reverse( ulresult );
+    return (int)slresult;
+}
+
+KOKKOS_FORCEINLINE_FUNCTION
+constexpr uint64_t bit_spread( const uint64_t mask, const int data )
+{
+    uint64_t rmask = binary_reverse( mask );
+    int dat = data;
+    uint64_t result = 0;
+    unsigned char lz{0};
+    while ( rmask )
+    {
+        lz = count_leading_zeros( rmask ) + 1;
+        result = result << lz | ( dat & 1 );
+        dat >>= 1, rmask <<= lz;
+    }
+    result = binary_reverse( result ) >> count_leading_zeros( mask );
+    return result;
+}
+//---------------------------------------------------------------------------//
+
+// Tile Hash Key (hash) <=> TileID
+template <typename Key, HashTypes HashT>
+struct TileID2HashKey;
+
+template <typename Key, HashTypes HashT>
+struct HashKey2TileID;
+
+// functions to compute hash key
+// can be rewriten in a recursive way
+template <typename Key>
+struct TileID2HashKey<Key, HashTypes::Naive>
+{
+    TileID2HashKey( std::initializer_list<int> tnum )
+    {
+        std::copy( tnum.begin(), tnum.end(), _tilenum.data() );
+    }
+    TileID2HashKey( int i, int j, int k )
+    {
+        _tilenum[0] = i;
+        _tilenum[1] = j;
+        _tilenum[2] = k;
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    constexpr auto operator()( int tile_i, int tile_j, int tile_k ) -> Key
+    {
+        return tile_i * _tilenum[1] * _tilenum[2] + tile_j * _tilenum[2] +
+               tile_k;
+    }
+
+  private:
+    Kokkos::Array<int, 3> _tilenum;
+};
+
+template <typename Key>
+struct TileID2HashKey<Key, HashTypes::Morton>
+{
+    TileID2HashKey( std::initializer_list<int> &&tnum ) {} // ugly
+    TileID2HashKey( int, int, int ) {}
+
+    enum : uint64_t
+    { // hand-coded now, can be improved by iterative func
+        page_kmask = ( 0x9249249249249249UL ),
+        page_jmask = ( 0x2492492492492492UL ),
+        page_imask = ( 0x4924924924924924UL )
+    };
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    constexpr auto operator()( int tile_i, int tile_j, int tile_k ) -> Key
+    {
+        return bit_spread( page_kmask, tile_k ) |
+               bit_spread( page_jmask, tile_j ) |
+               bit_spread( page_imask, tile_i );
+    }
+};
+
+// can be rewriten in a recursive way
+template <typename Key>
+struct HashKey2TileID<Key, HashTypes::Naive>
+{
+    HashKey2TileID( std::initializer_list<int> &&tnum )
+        : _tilenum()
+    {
+        std::copy( tnum.begin(), tnum.end(), _tilenum.data() );
+    }
+    HashKey2TileID( int i, int j, int k )
+    {
+        std::fill( _tilenum.data(), _tilenum.data() + 1, i );
+        std::fill( _tilenum.data() + 1, _tilenum.data() + 2, j );
+        std::fill( _tilenum.data() + 2, _tilenum.data() + 3, k );
```

Comment:
Same comment here.  Assign directly 
```suggestion
        _tilenum = { i, j, k };
```

Reply:
Sure, will fix these in the next commit